docs/src/literate-howto/threaded_assembly.jl

```diff
@@ -1,202 +1,316 @@
+# [OhMyThreads.jl](https://github.com/JuliaFolds2/OhMyThreads.jl) package. OhMyThreads
+# provides a macro based and a functional API. Here we use the macro based API because it is
+# slightly more convenient when using task local values since they can be defined with the
+# `@local` macro.
+#
+# !!! note "Schedulers and load balancing"
+#     OhMyThreads provides a number of different
+#     [schedulers](https://juliafolds2.github.io/OhMyThreads.jl/stable/refs/api/#Schedulers).
+#     In this example we use the `DynamicScheduler` (which is the default one). The
+#     `DynamicScheduler` will spawn `ntasks` tasks where each task will process a chunk of
+#     (roughly) equal number of cells (i.e. `length(color) ÷ ntasks`). This should be a good
+#     choice for this example because we expect all cells to take the same time to process
+#     and we don't need any load balancing.
+#
+#     For a different problem setup where some cells might take longer to process (perhaps
+#     they experience plastic deformation and we need to solve a local problem) we might
+#     benefit from load balancing. The `DynamicScheduler` can be used also for load
+#     balancing by specifiying `nchunks` or `chunksize`. However, the `DynamicScheduler`
+#     will always spawn `nchunks` tasks which can become costly since we are allocating
+#     scratch data for every task. To limit the number of tasks, while allowing for more
+#     than `ntasks` chunks, we can use the `GreedyScheduler` *with chunking*. For example,
+#     `scheduler = OhMyThreads.GreedyScheduler(; ntasks = ntasks, nchunks = 10 * ntasks)`
+#     will split the work into `10 * ntasks` chunks and spawn `ntasks` tasks to process
+#     them. Refer to the [OhMyThreads
+#     documentation](https://juliafolds2.github.io/OhMyThreads.jl/stable/) for details.
+
+using OhMyThreads, TaskLocalValues
+
+function assemble_global!(
+        K::SparseMatrixCSC, f::Vector, dh::DofHandler, colors,
+        cellvalues_tmp::CellValues; ntasks = Threads.nthreads()
+    )
+    ## Zero-out existing data in K and f
+    _ = start_assemble(K, f)
+    ## Body force and material stiffness
+    b = Vec{3}((0.0, 0.0, -1.0))
+    C = create_material_stiffness()
+    ## Loop over the colors
     for color in colors
-        ## Each color is safe to assemble threaded
-        Threads.@threads :static for i in 1:length(color)
-            assemble_cell!(scratches[Threads.threadid()], color[i], K, grid, dh, C, b)
+        ## Dynamic scheduler spawning `ntasks` tasks where each task will process a chunk of
+        ## (roughly) equal number of cells (`length(color) ÷ ntasks`).
+        scheduler = OhMyThreads.DynamicScheduler(; ntasks)
+        ## Parallelize the loop over the cells in this color
+        OhMyThreads.@tasks for cellidx in color
+            @set scheduler = scheduler
+            ## Obtain a task local scratch and unpack it
+            @local scratch = ScratchData(dh, K, f, cellvalues_tmp)
```

Comment:
Is the scratch data only created once per task? It does not look like it because it inside the loop, but maybe that is what the macro is for?

Reply:
https://juliafolds2.github.io/OhMyThreads.jl/stable/refs/api/#OhMyThreads.@local